source/server/software_enumerator.cpp

```diff
@@ -0,0 +1,92 @@
+#include "software_enumerator.h"
+
+using namespace std;
+
+namespace nidevice_grpc {
+
+SoftwareEnumerator::SoftwareEnumerator(SysCfgLibraryInterface* library)
+    : library_(library), syscfg_session_(nullptr)
+{
+}
+
+SoftwareEnumerator::~SoftwareEnumerator()
+{
+  // Caller is expected to close the syscfg session before calling destructor.
+}
+
+// Provides a list of installed software on a server under localhost. This internally uses the
+// "NI System Configuration API". If it is not currently installed, it can be downloaded from this page:
+// https://www.ni.com/en-in/support/downloads/drivers/download.system-configuration.html.
+::grpc::Status SoftwareEnumerator::enumerate_software(google::protobuf::RepeatedPtrField<SoftwareProperties>* software)
+{
+  NISysCfgStatus status = NISysCfg_OK;
+  NISysCfgSessionHandle session = NULL;
+  NISysCfgEnumSoftwareComponentHandle installedComps = NULL;
+  unsigned int numInstalledComps = 0;
+  char package_id[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+  char version[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+  char product_name[NISYSCFG_SIMPLE_STRING_LENGTH] = "";
+
+  try {
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library_->GetInstalledSoftwareComponents(session, NISysCfgIncludeItemsAllVisible, NISysCfgBoolTrue, &installedComps)))
+      {
+        if (NISysCfg_Succeeded(status = library_->ResetEnumeratorGetCount(installedComps, &numInstalledComps)) && numInstalledComps > 0)
+        {
+          while (NISysCfg_Succeeded(status) && (status = library_->NextComponentInfo(installedComps, package_id, version, product_name, NULL, NULL)) == NISysCfg_OK)
+          {
+            SoftwareProperties* properties = software->Add();
+            properties->set_package_id(package_id);
+            properties->set_version(version);
+            properties->set_product_name(product_name);
+          }
+        }
+        
+        library_->CloseHandle(installedComps);
+      }
+    }
+  }
+  catch (nidevice_grpc::NonDriverException& ex) {
+    return ex.GetStatus();
+  }
+
+  if (NISysCfg_Failed(status)) {
```

Comment:
Dan had a comment on this in Curtis's PR: https://github.com/ni/grpc-device/pull/772#discussion_r1004745616